### DISCUSSION TITLE
[alert/report] Solution for dashboards with multiple tabs

### DISCUSSION THREAD

**Discussion by zuzana-vej** [apache/superset#18475] [Ideas]
**Current situation / Problem statement**
Current Reporting solution only sends out the 1st dashboard tab. Dashboards can have multiple tabs as well as nested tabs (up to 3 levels of nested tabs are allowed). Within our user base, at least ~15% of dashboards have tabs. 

The workaround for users is either to put all the charts they want to receive via emailed on 1st tab, or create multiple dashboards just for the email report purpose. Some users do this, causing potentially clutter (many dashboards instead of one), making discoverability of the correct dashboard harder, and and making maintenance harder.

**User requirement:**
1. Ability to select which tab to send
2. Ability to send multiple tabs in a single email

**Rejected solutions**
1. Send all tabs; This can have negative performance implications; some dashboards are large (many tabs) and some dashboard producers create tab just for old / deprecated charts, and those should never be send out.

**Proposed solution**
Enable user to select which tabs to be send out; User can select more than one tab. The email will contain multiple screenshots, in the order of most top level / most left tab as if user was browsing the dashboard. 

**Discussion points**
1. Should there be an upper limit to how many tabs user can select to send within a single email?
2. UI/UX

**Comment by zuzana-vej** [comment]
Update: We are planning to prioritize this feature at some point late Q2/earlyQ3, but keeping this GH# open as a space for discussion (to enable this tab specific discussion separately from the #13954

**Comment by zuzana-vej** [comment]
**Few data points from Airbnb Email Report users:**
~50 users; 11 responded a survey; 65% of users would use this feature immediately; 75% plan to only send multiple top level tab; 25% would use it for specifying a nested tabs to be send out.
Users would like to send different tabs to different recipients (= can be satisfied as multiple report configurations).

**Comment by zuzana-vej** [comment]
Would love some design inputs on how to enable users to select multiple tabs which can then be attached to a single email report. Options we thought about:

1. multiple drop down boxes 
-- OK for one tab, but might be complicated for multiple tabs)

2. tabs combinations can be fully spelled out in a single dropdown and user can select 1+ items from the dropdown
Example (sample dashboard):
<img width="671" alt="Screen Shot 2021-10-18 at 1 17 42 PM" src="https://user-images.githubusercontent.com/61221714/138030398-22b88ac3-f0e5-43e3-9dde-72d5f4270f99.png">
Dropdown would have values:

> 1. Summary 
> 2. Deep Dive | Summary
> 2. Deep Dive | Page Load Duration
> 2. Deep Dive | Usage Details
> 2. Deep Dive | Error Details
> 2. Deep Dive | Cache and Query Time
>  

Would love some inputs from @jess-dillard / @mihir174  / @yousoph 
(Cc @graceguo-supercat, @m-ajay)

**Comment by ghost** [comment]
Have you considered checkboxes? Nested tabs could be represented as indented checkboxes, where the top-level checkbox could control all beneath it.

**Comment by zuzana-vej** [comment]
There are two ways how user can get to Email report modal configuration - user has 2 ways to enter into the "editing report" flow.

**1. From dashboard:** 
- User is on dashboard and clicks on the icon in top right corner.  User might are viewing dashbord in this moment and are seeing **specific tabs combination** when they click on the report icon. 
- The report will get configured for the **current tabs** user is viewing.  
- User will see an information `configuring report for currently selected tabs` --> the proposal is to add this text above "REPORT NAME" textbox on the modal
- There is no need to add tab selector into the modal, because user will be configuring for the currently visible tabs
<img width="379" alt="modal-1" src="https://user-images.githubusercontent.com/61221714/140182897-0071467c-1358-49fb-8873-f35c70e5f7bf.png">

**2. From "Settings > Alerts&Reports"**
- User can enter the report configuration from "Settings > Alerts&Reports" and click "+Alert" button. 
- The complex new report modal shows up and in this modal, user will be able to select the tabs to be sent out
<img width="1155" alt="modal-2" src="https://user-images.githubusercontent.com/61221714/140181734-0bbd23d2-286b-4cbe-b724-e0c361c2188b.png">

The experience to select tabs will be like tree style check boxes to display the tabs hierarchy, similarly like filter scopes:
<img width="139" alt="Screen Shot 2021-11-03 at 10 16 08 AM" src="https://user-images.githubusercontent.com/61221714/140182268-77b9bc0d-047a-452c-888e-2345008f214b.png">
- User can select any number of child tabs
- Is user selects a top level tab (or a tab which still has child tabs), by default only the child top level tab will get selected. (reason (1) mimic experience of viewing dashboard; (2) to prevent too many screenshots of dashboard unless needed as indicated by user)
- If user wants to send all child level tabs of a specific top level tab, they can manually check all the child checkboxes.

**Limits:**
Currently no plans to limit how many tabs user can select; if there are performance issues, we will revisit this area and add a warning or config variable to limit the tabs user can send.


Please share if you have any feedback to the plan. @jess-dillard @yousoph

**Comment by graceguo-supercat** [comment]
I agree @zuzana-vej's proposal, just 1 small suggestion:
For case 1 **From dashboard:**
instead of enforcing report on the current selected tab, give user an option: report default dashboard (by default), or report current selected tab. The UI could be similar to filter component scope configuration:
<img width="423" alt="Screen Shot 2021-11-04 at 10 11 44 AM" src="https://user-images.githubusercontent.com/27990562/140391976-82c2e54d-a4c0-451c-b2ec-a13cac6cffd2.png">
This behavior will be consistent with current dashboard report behavior:  by default, report the first tab (and its first children tabs).

Similarly for case 2 **From "Settings > Alerts&Reports"**
Also add an option: report default dashboard (by default), or select a tab. If user choose select a dashboard tab, then show the dashboard tabs tree. 
cc @zuzana-vej

**Comment by yousoph** [comment]
This is great! The proposal makes sense to me, @zuzana-vej and I like @graceguo-supercat 's suggestion of giving the user more information and choice when setting up a report. 

I'd also suggest allowing users from the dashboard to select multiple tabs for their report. Maybe by default the report is for the current tab but they can select the default dashboard or additional tabs as well?

**Comment by zuzana-vej** [comment]
@yousoph thanks for the suggestions; it can be done. 

Different question - what was the initial thinking behind having tow different modals based on user entry point? The one from Settings is much more details which is why we though the multiple tabs is a detail setting and therefore can be on the details modal.  

I am unsure how many users know that exists and whether a note on the modal they get when entering from Dashboard, saying something like `More detailed settings, like multiple owners, conditional alert, or multiple tabs, is available from Settings > Alerts & Reports.`

**Comment by graceguo-supercat** [comment]
Hi @yousoph I also prefer Zuzana's proposal: set default or selected tab as _selected report tab_ inside a dashboard's report menu, and show tab selector in the report detail menu. 

When setup report schedule, I feel people want to make sure they are sending out the right tab, so they want to verify it in the dashboard during the setup schedule. 

- In dashboard, users are easy to navigate to the right place, and set. So inside dashboard, create a new selector and ask user to choose a tab to report, seems not necessary.

- In report detail page, preview dashboard is a little harder, so ppl probably to open a dashboard in other browser and check it. But probably we don't have better way for user to preview selected tab inside report menu.

**Comment by agusfigueroa-htg** [comment]
Hi everyone! Looks like this is a common issue around, does anyone have news on this/knows about any related PR? 
Thanks 🙏

**Reply by zuzana-vej** [reply]
as far as I know it's WIP. @graceguo-supercat can share more.

**Comment by sujiplr** [comment]
Awesome discussion and solution. 
I'm having a suggestion, Is it possible to have the option to send screenshot as single email or multiple emails(each one for a single tab). In most of the cases, we will be keeping different set of data cuts in the different tabs, hence for some of the cases it will be more readable when it comes as multiple emails. Many users ( management people ) nowadays access emails via Mobile/Tabs, so if we give it as different emails for each tab, will give more readability rather than giving all in one email, we can give this as an option while configuring repots.

**Reply by zuzana-vej** [reply]
This would be possible if you setup multiple report configurations - one for each tab you want to send (and can define same or different recipients).

**Comment by giovannipapini-agilelab** [comment]
Hi, are there any updates about current status of this feature?

**Comment by webdev-creator** [comment]
Hi, It's not a big deal. Actually, you can customize this feature to add separate multiple tabs into images. The images will be attached to the email or slack.

**Reply by agusfigueroa-htg** [reply]
Is this already available?

**Reply by yammt** [reply]
@agusfigueroa-htg I've customized it for my client.

**Comment by sfirke** [comment]
A PR was merged in June 2024 that implemented the ability to choose which tab to send: https://github.com/apache/superset/pull/29096
I don't believe it's currently possible to send multiple tabs in a single alert/report.
